### PR TITLE
Fix Delivered and Processed message handling

### PR DESCRIPTION
**Pull request by hackaugusto**
Partially-implements: #3254 

- Processed messages should not be confirmed by Delivered messages
- Processed messages are not retriable, should not be added to the channel queue
- Messages in the retriable queue should be cleared with a Processed and not a Delivered

### CODE REVIEW

File: raiden/transfer/channel.py [hunk 1]
```diff
@@ -2010,7 +2010,6 @@ def handle_receive_withdraw_confirmation(
             SendProcessed(
                 recipient=channel_state.partner_state.address,
                 message_identifier=withdraw.message_identifier,
-                canonical_identifier=CANONICAL_IDENTIFIER_UNORDERED_QUEUE,
```
Comment (palango) [inline raiden/transfer/channel.py]: Nice that we can remove these!

File: raiden/tests/unit/storage/test_serialization.py [hunk 2]
```diff
@@ -140,7 +140,10 @@ def test_restore_queueids_to_queues(chain_state, netting_channel_state):
     chain_state.queueids_to_queues[queue_identifier] = messages
 
     serialized_chain_state = JSONSerializer.serialize(chain_state)
-
     deserialized_chain_state = JSONSerializer.deserialize(serialized_chain_state)
 
+    for f in dir(chain_state):
```
Comment (palango) [inline raiden/tests/unit/storage/test_serialization.py]: Why is this necessary?

File: raiden/transfer/mediated_transfer/target.py [hunk 3]
```diff
@@ -243,13 +241,7 @@ def handle_unlock(
...
         )
 
-        send_processed = SendProcessed(
-            recipient=balance_proof_sender,
-            message_identifier=state_change.message_identifier,
-            canonical_identifier=CANONICAL_IDENTIFIER_UNORDERED_QUEUE,
-        )
-
-        events.extend([payment_received_success, unlock_success, send_processed])
+        events.extend([payment_received_success, unlock_success])
```
Comment (palango) [inline raiden/transfer/mediated_transfer/target.py]: Why don't we send the `Processed` here any more?

File: raiden/tests/unit/storage/test_serialization.py (hunk 2 above)
Comment (hackaugusto) [reply raiden/tests/unit/storage/test_serialization.py]: oops, that is not necessary, I added it while debugging.

File: raiden/transfer/mediated_transfer/target.py (hunk 3 above)
Comment (hackaugusto) [reply raiden/transfer/mediated_transfer/target.py]: that is what the PR fix, only the Delivered is necessary.

File: raiden/messages/healthcheck.py [hunk 4]
```diff
@@ -33,7 +33,7 @@ def _data_to_sign(self) -> bytes:
 
 
 @dataclass(repr=False, eq=False)
-class Pong(SignedMessage):
+class Pong(SignedRetriableMessage):
```
Comment (rakanalh) [inline raiden/messages/healthcheck.py]: **Note:** Ping/Pong are currently only used in tests. 

The reason why this change came to be is that previously, the transport relied on `Processed -> Delivered` cycle to test certain aspects of the itself. However, due to the removal of `Delivered` confirmation to `Processed`, the tests became invalid. The way to move forward was to be able to call `send_async` for `Ping` and `Pong` so that these tests could be restored to a passable state. 

The original `MessageHandler` used for Raiden is unable to process `Ping`/`Pong` that's why the `test_matrix_transport.py` `MessageHandler` object was modified to handle `Ping` messages by replying to `Pong`.

I'll wait for the confirmation that the fix works once the build for the current commits has completed.

File: raiden/network/transport/matrix/transport.py [hunk 5]
```diff
@@ -992,17 +991,16 @@ def _handle_sync_messages(self, sync_messages: MatrixSyncMessages) -> bool:
                 for text in room_messages:
                     all_messages.extend(self._handle_text(room, text))
 
-        # Remove this #3254
         for message in all_messages:
-            if isinstance(message, (Processed, SignedRetrieableMessage)) and message.sender:
```
Comment (rakanalh) [inline raiden/network/transport/matrix/transport.py]: This was missed. I excluded `Processed` from the `Delivered` confirmation as the PR intended.

File: raiden/transfer/node.py [hunk 6]
```diff
@@ -564,11 +560,9 @@ def handle_contract_receive_channel_closed(
 
 
 def handle_receive_delivered(
-    chain_state: ChainState, state_change: ReceiveDelivered
+    chain_state: ChainState, state_change: ReceiveDelivered  # pylint: disable=unused-argument
 ) -> TransitionResult[ChainState]:
     """ Check if the "Delivered" message exists in the global queue and delete if found."""
-    queueid = QueueIdentifier(state_change.sender, CANONICAL_IDENTIFIER_UNORDERED_QUEUE)
-    inplace_delete_message_queue(chain_state, state_change, queueid)
```
Comment (rakanalh) [inline raiden/transfer/node.py]: This basically means that we can entirely remove `Delivered` IMO. Need to verify but i think it's possible.

@hackaugusto thoughts?

**Comment by rakanalh** [comment]
`Delivered` no longer confirms `Processed` and `Processed` is made as a non-retriable message... this means that if the queue had a `Processed` message that was not yet sent and Raiden restarts, the message is lost. As a result, the partner might send a `LockedTransfer` message which is going to received by us as `ActionInitTarget` and we would fail with `EventUnlockClaimFailed` because the message was already received and processed but we failed to inform the partner about the fact that it's processed. Which makes me think about the restartability aspect of alot of other places where this might happen.

The solution i can think of for this is to accept messages with an old nonce and reply with a processed if the current nonce is larger than the one received in a given message.

This means that `is_valid_*` functions in the channel state machine need to be adjusted for this. An example of the modification for `ActionInitTarget` can be seen in `is_valid_lockedtransfer` called by `handle_receive_lockedtransfer`.

@hackaugusto what do you think?

**Comment by Dominik1999** [comment]
@ulope please make sure to update the Raiden spec accordingly or even better get help from someone to parallelize that

**Comment by ulope** [comment]
@hackaugusto and I looked more into this today and found that the proposed mechanism of re-sending `Processed` messages for retried old messages isn't straight forward.

The problem occurs for messages that don't contain a nonce (either explicitly or via a balance proof). In those cases we would need to infer the validity of the message through some other means.

Possibilities of that:
- Keep a set of all previously received valid messages (not desirable since it will grow indefinitely)
- Perform a database lookup (performance impact would need investigation)

Therefore we were thinking of postponing this PR until after the Alderaan release.
The downside of that would be the need to implement a backwards compatibility once we do implement this PR (or another implementation of #3254)

File: raiden/transfer/mediated_transfer/target.py (hunk 3 above)
Comment (konradkonrad) [reply raiden/transfer/mediated_transfer/target.py:246]: this seems to be a valid bug. Unlock needs to be answered by a `Processed`.
